### PR TITLE
Remove connect SocketPermissions from core

### PR DESCRIPTION
**Pull request by Tim-Brooks**
This permission is relegated to these modules/plugins:
- transport-netty4 module
- reindex module
- repository-url module
- discovery-azure-classic plugin
- discovery-ec2 plugin
- discovery-gce plugin
- repository-azure plugin
- repository-gcs plugin
- repository-hdfs plugin
- repository-s3 plugin

And for tests:
- mocksocket jar
- rest client
- httpcore-nio jar
- httpasyncclient jar

### CODE REVIEW

**Comment by tbrooks8** [comment]
A couple notes:

1. This depends on #22793
2. It's a little messy that I need to give permissions to rest client, httpcore-nio, and httpasyncclient in tests. But this is necessary for the rest integration tests. The `doPrivileged` block is at the rest client level. So the http libraries (which are called by rest) also need the permissions.
3. Due to the way tests permission are applied, some issues can be obscured. For ex: `reindex` needs `connect`. But even if I did not give it `connect`, the tests would still pass. This is because it depends on `transport-netty4` which has `connect`. And the way privileges are applied in tests, `reindex` dependencies get any permissions `transport-netty4` when the `transport-netty4` security policy is applied in `reindex` tests. I'm not sure if there is a clear way around this without changing the build significantly.

File: core/src/main/resources/org/elasticsearch/bootstrap/security.policy [hunk 1]
```diff
@@ -55,8 +55,8 @@ grant {
   // third party code, to safeguard these against unprivileged code like scripts.
   permission org.elasticsearch.SpecialPermission;
 
-  // Allow connecting to the internet anywhere
-  permission java.net.SocketPermission "*", "connect,resolve";
+  // Allow host/ip name service lookups
```
Comment (s1monw) [inline core/src/main/resources/org/elasticsearch/bootstrap/security.policy]: w00t

**Review by s1monw — APPROVED** [review]
LGTM

File: core/src/main/resources/org/elasticsearch/bootstrap/test-framework.policy [hunk 2]
```diff
@@ -64,6 +64,22 @@ grant codeBase "${codebase.junit-4.11.jar}" {
...
+  permission java.net.SocketPermission "*", "accept,connect";
+};
+
+
+grant codeBase "${codebase.rest-6.0.0-alpha1-SNAPSHOT.jar}" {
+  // rest makes socket connections for rest tests
+  permission java.net.SocketPermission "*", "connect";
+};
+
+grant codeBase "${codebase.httpcore-nio-4.4.5.jar}" {
```
Comment (jasontedor) [inline core/src/main/resources/org/elasticsearch/bootstrap/test-framework.policy]: Can you add comments to the places where these dependencies are defined that if the version is changed then the version needs to be updated here too? (This comment applies to all of the dependencies for which we have to supply these permissions.)

**Review by jasontedor — CHANGES_REQUESTED** [review]
I left one comment.

File: core/src/main/resources/org/elasticsearch/bootstrap/test-framework.policy [hunk 3]
```diff
@@ -64,6 +64,22 @@ grant codeBase "${codebase.junit-4.11.jar}" {
...
 
 grant codeBase "${codebase.mocksocket-1.1.jar}" {
-  // mocksocket accepts socket connections
-  permission java.net.SocketPermission "*", "accept";
+  // mocksocket makes and accepts socket connections
+  permission java.net.SocketPermission "*", "accept,connect";
+};
+
+
+grant codeBase "${codebase.rest-6.0.0-alpha1-SNAPSHOT.jar}" {
```
Comment (jasontedor) [inline core/src/main/resources/org/elasticsearch/bootstrap/test-framework.policy]: Does this one deserve a comment too, above the `elasticsearch` version in version.properties?

**Review by jasontedor — CHANGES_REQUESTED** [review]
I left one more comment.

**Review by jasontedor — APPROVED** [review]
LGTM.

**Pull request merged**